### PR TITLE
Add PropertyUsedBeforeDeclaration rule

### LINKED ISSUE #749: Check if property is used before assignment
For example in this case:
```    
class Foo {
    val number
        get() = if (isValid) 1 else 0
    val list = listOf(number)
    val isValid = true
}
println(Foo().list)
```
`0` will be printed because `isValid` is `false` when getter is invoked first time. If you move `val isValid = true` at least one line above, `1` will be printed.

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #749

### CODE REVIEW

**Review by cortinico — APPROVED** [review]
LGTM 👍

File: detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PropertyUsedBeforeDeclarationSpec.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import io.gitlab.arturbosch.detekt.test.compileAndLintWithContext
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.junit.jupiter.api.Test
+
+@KotlinCoreEnvironmentTest
+class PropertyUsedBeforeDeclarationSpec(private val env: KotlinCoreEnvironment) {
+    private val subject = PropertyUsedBeforeDeclaration()
+
+    @Test
+    fun `used before declaration in getter`() {
```
Comment (schalkms) [inline detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PropertyUsedBeforeDeclarationSpec.kt:14]: I'm wondering whether other scenarios exist where a test case would be beneficial.
However, I couldn't come up with any scenarios, which you haven't already added.

**Review by schalkms — APPROVED** [review]
Awesome! Thank you very much for this nice addition to detekt. 🙏

File: detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PropertyUsedBeforeDeclaration.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            val descriptor = bindingContext[BindingContext.DECLARATION_TO_DESCRIPTOR, it] ?: return@mapNotNull null
+            name to descriptor
+        }.toMap()
+
+        val declaredProperties = mutableSetOf<DeclarationDescriptor>()
+
+        classMembers.forEach { member ->
+            member.forEachDescendantOfType<KtNameReferenceExpression> {
+                val property = allProperties[it.text]
+                if (property != null && property !in declaredProperties && property == it.descriptor()) {
```
[the remark this answers is not in the document]
Comment (t-kameyama) [reply detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PropertyUsedBeforeDeclaration.kt:78]: `variable shadowing` test will fail without the null check.
The type name is also a KtNameReferenceExpression, so we can't omit it.

[the remark this answers is not in the document]
Comment (schalkms) [reply detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PropertyUsedBeforeDeclaration.kt:78]: Thanks - resolved!

**Pull request merged**